fluss-common/src/main/java/com/alibaba/fluss/row/encode/paimon/PaimonBinaryRowWriter.java

```diff
@@ -0,0 +1,420 @@
+ * limitations under the License.
+ */
+
+package com.alibaba.fluss.row.encode.paimon;
+
+import com.alibaba.fluss.memory.MemorySegment;
+import com.alibaba.fluss.record.RowKind;
+import com.alibaba.fluss.row.BinarySegmentUtils;
+import com.alibaba.fluss.row.BinaryString;
+import com.alibaba.fluss.row.Decimal;
+import com.alibaba.fluss.row.InternalRow;
+import com.alibaba.fluss.row.TimestampLtz;
+import com.alibaba.fluss.row.TimestampNtz;
+import com.alibaba.fluss.types.DataType;
+import com.alibaba.fluss.utils.UnsafeUtils;
+
+import java.io.Serializable;
+import java.nio.charset.StandardCharsets;
+import java.util.Arrays;
+
+import static com.alibaba.fluss.types.DataTypeChecks.getPrecision;
+
+/**
+ * A writer to encode Fluss's {@link InternalRow} using Paimon's BinaryRow encoding way.
+ *
+ * <p>The logic is almost copied from Paimon's BinaryRowWriter.
+ */
+class PaimonBinaryRowWriter {
+
+    private static final int HEADER_SIZE_IN_BITS = 8;
+    private static final int MAX_FIX_PART_DATA_SIZE = 7;
+
+    private final int nullBitsSizeInBytes;
+    private final int fixedSize;
+    private byte[] buffer;
+    private MemorySegment segment;
+
+    private int cursor;
+
+    public PaimonBinaryRowWriter(int arity) {
+        this.nullBitsSizeInBytes = calculateBitSetWidthInBytes(arity);
+        this.fixedSize = getFixedLengthPartSize(nullBitsSizeInBytes, arity);
+        this.cursor = fixedSize;
+        setBuffer(new byte[fixedSize]);
+    }
+
+    public void reset() {
+        this.cursor = fixedSize;
+        for (int i = 0; i < nullBitsSizeInBytes; i += 8) {
+            UnsafeUtils.putLong(buffer, i, 0L);
+        }
+    }
+
+    public byte[] toBytes() {
+        byte[] bytes = new byte[cursor];
+        System.arraycopy(buffer, 0, bytes, 0, cursor);
+        return bytes;
+    }
+
+    public void setNullAt(int pos) {
+        setNullBit(pos);
+        UnsafeUtils.putLong(buffer, getFieldOffset(pos), 0L);
+    }
+
+    private void setNullBit(int pos) {
+        UnsafeUtils.bitSet(buffer, 0, pos + HEADER_SIZE_IN_BITS);
+    }
+
+    public void writeRowKind(RowKind kind) {
+        // Fluss has APPEND_ONLY rowKind, so minus 1 to align with paimon
+        byte kindByte = (byte) (kind.toByteValue() - 1);
+        UnsafeUtils.putByte(buffer, 0, kindByte);
+    }
+
+    public void writeBoolean(int pos, boolean value) {
+        UnsafeUtils.putBoolean(buffer, getFieldOffset(pos), value);
+    }
+
+    public void writeByte(int pos, byte value) {
+        UnsafeUtils.putByte(buffer, getFieldOffset(pos), value);
+    }
+
+    public void writeShort(int pos, short value) {
+        UnsafeUtils.putShort(buffer, getFieldOffset(pos), value);
+    }
+
+    public void writeInt(int pos, int value) {
+        UnsafeUtils.putInt(buffer, getFieldOffset(pos), value);
+    }
+
+    public void writeLong(int pos, long value) {
+        UnsafeUtils.putLong(buffer, getFieldOffset(pos), value);
```

Comment:
I recommend using Fluss's `MemorySegment#putLong` to write data. The issue with `Unsafe#putLong` is that it relies on the native byte order of the machine, which can lead to data corruption if the writer and reader operate on systems with different byte orders. Storage systems should always use a consistent, predefined byte order rather than depending on the machine's native order. I think this is a design flaw in Paimon. 

Since nearly all modern architectures (e.g., x86, ARM) use little-endian, we can safely assume Paimon uses little-endian as well. Using Fluss's `MemorySegment#putLong`, which guarantees little-endian encoding, ensures compatibility and avoids potential data corruption.

Reply:
Thanks..Agree with you.. At first, I also want to use `MemorySegment#putLong`, but find Paimon use  native byte order of the machine, so I use `Unsafe#putLong` to align with Paimon.